File: weather-library/src/main/java/org/rcgonzalezf/weather/openweather/network/OpenWeatherApiRequestParameters.java
package org.rcgonzalezf.weather.openweather.network;

import org.rcgonzalezf.weather.common.network.RequestParameters;

public class OpenWeatherApiRequestParameters implements RequestParameters {

  private String lat;
  private String lon;
  private String cityName;

  public String getLat() {
    return lat;
  }

  public void setLat(String lat) {
    this.lat = lat;
  }

  public String getLon() {
    return lon;
  }

  public void setLon(String lon) {
    this.lon = lon;
  }

  String getCityName() {
    return cityName;
  }

  private void setCityName(String cityName) {
    this.cityName = cityName;
  }

  public static class OpenWeatherApiRequestBuilder {

    public static final String LON = "lon";
    public static final String LAT = "lat";
    public static final String TYPE = "type";
    static final String LIKE = "like";
    private final OpenWeatherApiRequestParameters mOpenWeatherApiRequestParameters;


    public OpenWeatherApiRequestBuilder() {
      mOpenWeatherApiRequestParameters = new OpenWeatherApiRequestParameters();
    }

    public OpenWeatherApiRequestParameters build() {
      return mOpenWeatherApiRequestParameters;
    }

    private OpenWeatherApiRequestBuilder withLat(Double lat) {
      mOpenWeatherApiRequestParameters.setLat(String.valueOf(lat));
      return this;
    }

    private OpenWeatherApiRequestBuilder withLon(Double lon) {
      mOpenWeatherApiRequestParameters.setLon(String.valueOf(lon));
      return this;
    }

    public OpenWeatherApiRequestBuilder withLatLon(Double lat, Double lon) {
      return withLat(lat).withLon(lon);
    }

    public OpenWeatherApiRequestBuilder withCityName(String cityName) {
      mOpenWeatherApiRequestParameters.setCityName(cityName);
      return this;
    }
  }
}

File: weather-library/src/main/java/org/rcgonzalezf/weather/openweather/network/OpenWeatherApiRequestParameters.kt
package org.rcgonzalezf.weather.openweather.network

import org.rcgonzalezf.weather.common.network.RequestParameters
open // for Mockito
class OpenWeatherApiRequestParameters : RequestParameters {
    var lat: String? = null
    var lon: String? = null
    var cityName: String? = null
        private set

    private fun setCityName(cityName: String) {
        this.cityName = cityName
    }

    class OpenWeatherApiRequestBuilder {
        private val openWeatherApiRequestParameters: OpenWeatherApiRequestParameters
                = OpenWeatherApiRequestParameters()

        companion object {
            const val LON = "lon"
            const val LAT = "lat"
            const val TYPE = "type"
            const val LIKE = "like"
        }

        fun build(): OpenWeatherApiRequestParameters {
            return openWeatherApiRequestParameters
        }

        private fun withLat(lat: Double): OpenWeatherApiRequestBuilder {
            openWeatherApiRequestParameters.lat = lat.toString()
            return this
        }

        private fun withLon(lon: Double): OpenWeatherApiRequestBuilder {
            openWeatherApiRequestParameters.lon = lon.toString()
            return this
        }

        fun withLatLon(lat: Double, lon: Double): OpenWeatherApiRequestBuilder {
            return withLat(lat).withLon(lon)
        }

        fun withCityName(cityName: String): OpenWeatherApiRequestBuilder {
            openWeatherApiRequestParameters.setCityName(cityName)
            return this
        }
    }
}
